In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

df = pd.read_csv("../data/Traffic_Crashes_Resulting_in_Fatality.csv")

df.head()


unique_id case_id_fkey   latitude   longitude  collision_year  death_date  \
0          1    140236301  37.710409 -122.404226            2014  2014/03/20   
1          2    140755533  37.725476 -122.394243            2014  2014/09/08   
2          4    140365546  37.748255 -122.413669            2014  2014/05/03   
3         16    150562049  37.777300 -122.419694            2015  2015/06/30   
4         17    140104811  37.778251 -122.419883            2014  2014/02/06   

  death_time          death_datetime collision_date collision_time  ...  \
0   11:21:00  2014/03/20 11:21:00 AM     2014/03/20            NaN  ...   
1   16:38:00  2014/09/08 04:38:00 PM     2014/09/08       05:10:00  ...   
2   17:20:00  2014/05/03 05:20:00 PM     2014/05/03       02:24:00  ...   
3   06:00:00  2015/06/30 06:00:00 AM     2015/06/28       03:52:00  ...   
4   10:20:00  2014/02/06 10:20:00 AM     2014/02/05       02:26:00  ...   

  in_coc_2018 publish  on_vz_hin_2022 in_epa_2021  \
0       False    True            True       False   
1        True    True            True        True   
2       False    True            True       False   
3       False    True            True        True   
4       False    True            True        True   

                                 point  analysis_neighborhood  \
0  POINT (-122.404226037 37.710409217)  Bayview Hunters Point   
1   POINT (-122.394243493 37.72547565)  Bayview Hunters Point   
2  POINT (-122.413668844 37.748255329)                Mission   
3  POINT (-122.419693566 37.777299856)             Tenderloin   
4  POINT (-122.419883231 37.778251017)           Hayes Valley   

  supervisor_district  police_district              data_as_of  \
0                10.0        INGLESIDE  2024/09/18 12:00:00 AM   
1                10.0          BAYVIEW  2024/09/18 12:00:00 AM   
2                 9.0          MISSION  2024/09/18 12:00:00 AM   
3                 5.0         NORTHERN  2024/09/18 12:00:00 AM   
4                 5.0         NORTHERN  2024/09/18 12:00:00 AM   

           data_loaded_at  
0  2025/11/06 10:51:39 AM  
1  2025/11/06 10:51:39 AM  
2  2025/11/06 10:51:39 AM  
3  2025/11/06 10:51:39 AM  
4  2025/11/06 10:51:39 AM  

[5 rows x 28 columns]

In [2]:

df.shape


(345, 28)

In [3]:
df.columns


Index(['unique_id', 'case_id_fkey', 'latitude', 'longitude', 'collision_year',
       'death_date', 'death_time', 'death_datetime', 'collision_date',
       'collision_time', 'collision_datetime', 'location', 'age', 'sex',
       'deceased', 'collision_type', 'street_type', 'on_vz_hin_2017',
       'in_coc_2018', 'publish', 'on_vz_hin_2022', 'in_epa_2021', 'point',
       'analysis_neighborhood', 'supervisor_district', 'police_district',
       'data_as_of', 'data_loaded_at'],
      dtype='object')

In [4]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 345 entries, 0 to 344
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   unique_id              345 non-null    int64  
 1   case_id_fkey           338 non-null    object 
 2   latitude               345 non-null    float64
 3   longitude              345 non-null    float64
 4   collision_year         345 non-null    int64  
 5   death_date             345 non-null    object 
 6   death_time             311 non-null    object 
 7   death_datetime         345 non-null    object 
 8   collision_date         345 non-null    object 
 9   collision_time         320 non-null    object 
 10  collision_datetime     345 non-null    object 
 11  location               345 non-null    object 
 12  age                    344 non-null    float64
 13  sex                    345 non-null    object 
 14  deceased               345 non-null    object 
 15  collis

In [6]:
df.select_dtypes(include='number').describe()


unique_id    latitude   longitude  collision_year         age  \
count  3.450000e+02  345.000000  345.000000      345.000000  344.000000   
mean   1.319810e+07   37.762547 -122.428598     2019.594203   52.531977   
std    1.090101e+08    0.027374    0.030388        3.463761   20.924663   
min    1.000000e+00   37.708628 -122.513360     2014.000000    0.000000   
25%    1.140000e+02   37.741162 -122.442648     2016.000000   34.000000   
50%    2.750000e+02   37.769586 -122.419714     2020.000000   53.000000   
75%    2.023001e+07   37.783785 -122.407587     2023.000000   68.000000   
max    2.024040e+09   37.823175 -122.370075     2025.000000   95.000000   

       supervisor_district  
count           344.000000  
mean              6.284884  
std               3.030497  
min               1.000000  
25%               4.000000  
50%               6.000000  
75%               9.000000  
max              11.000000

# Data Overview

In this notebook, I (Zhanna) will do:

- Data cleaning for the traffic fatality dataset.
- Part 1: Death trends (by year, month, and time of day).

From the initial inspection:

- The dataset has `345` rows and `28` columns.
- Important time-related columns include `collision_date`, `collision_time`, and `collision_datetime`.
- I will use `collision_datetime` to define when the fatal crash occurred.


In [7]:
# Convert date and datetime columns to proper datetime format

date_columns = [
    'death_date', 'death_datetime',
    'collision_date', 'collision_datetime',
    'data_as_of', 'data_loaded_at'
]

for col in date_columns:
    df[col] = pd.to_datetime(df[col], errors='coerce')

df.head()


/tmp/ipykernel_826/3247978513.py:10: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[col] = pd.to_datetime(df[col], errors='coerce')
/tmp/ipykernel_826/3247978513.py:10: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df[col] = pd.to_datetime(df[col], errors='coerce')


unique_id case_id_fkey   latitude   longitude  collision_year death_date  \
0          1    140236301  37.710409 -122.404226            2014 2014-03-20   
1          2    140755533  37.725476 -122.394243            2014 2014-09-08   
2          4    140365546  37.748255 -122.413669            2014 2014-05-03   
3         16    150562049  37.777300 -122.419694            2015 2015-06-30   
4         17    140104811  37.778251 -122.419883            2014 2014-02-06   

  death_time      death_datetime collision_date collision_time  ...  \
0   11:21:00 2014-03-20 11:21:00     2014-03-20            NaN  ...   
1   16:38:00 2014-09-08 16:38:00     2014-09-08       05:10:00  ...   
2   17:20:00 2014-05-03 17:20:00     2014-05-03       02:24:00  ...   
3   06:00:00 2015-06-30 06:00:00     2015-06-28       03:52:00  ...   
4   10:20:00 2014-02-06 10:20:00     2014-02-05       02:26:00  ...   

  in_coc_2018 publish  on_vz_hin_2022 in_epa_2021  \
0       False    True            True       False   
1        True    True            True        True   
2       False    True            True       False   
3       False    True            True        True   
4       False    True            True        True   

                                 point  analysis_neighborhood  \
0  POINT (-122.404226037 37.710409217)  Bayview Hunters Point   
1   POINT (-122.394243493 37.72547565)  Bayview Hunters Point   
2  POINT (-122.413668844 37.748255329)                Mission   
3  POINT (-122.419693566 37.777299856)             Tenderloin   
4  POINT (-122.419883231 37.778251017)           Hayes Valley   

  supervisor_district  police_district  data_as_of      data_loaded_at  
0                10.0        INGLESIDE  2024-09-18 2025-11-06 10:51:39  
1                10.0          BAYVIEW  2024-09-18 2025-11-06 10:51:39  
2                 9.0          MISSION  2024-09-18 2025-11-06 10:51:39  
3                 5.0         NORTHERN  2024-09-18 2025-11-06 10:51:39  
4                 5.0         NORTHERN  2024-09-18 2025-11-06 10:51:39  

[5 rows x 28 columns]

In [8]:
# Extract useful time components from collision_datetime

df['year'] = df['collision_datetime'].dt.year
df['month'] = df['collision_datetime'].dt.month
df['day'] = df['collision_datetime'].dt.day
df['hour'] = df['collision_datetime'].dt.hour
df['month_name'] = df['collision_datetime'].dt.month_name()

df[['year', 'month', 'day', 'hour']].head()


year  month  day  hour
0  2014      3   20     0
1  2014      9    8     5
2  2014      5    3     2
3  2015      6   28     3
4  2014      2    5     2

In [9]:
# Define categorical time of day

def time_of_day(h):
    if pd.isna(h):
        return "Unknown"
    if 0 <= h < 6:
        return "Night (00–05)"
    elif 6 <= h < 12:
        return "Morning (06–11)"
    elif 12 <= h < 18:
        return "Afternoon (12–17)"
    else:
        return "Evening (18–23)"

df['time_of_day'] = df['hour'].apply(time_of_day)

df['time_of_day'].value_counts()


time_of_day
Night (00–05)        99
Evening (18–23)      92
Afternoon (12–17)    91
Morning (06–11)      63
Name: count, dtype: int64

In [10]:
df = df[(df['year'] >= 2014) & (df['year'] <= 2025)]
df['year'].value_counts().sort_index()


year
2014    30
2015    31
2016    29
2017    20
2018    23
2019    29
2020    30
2021    27
2022    39
2023    26
2024    42
2025    19
Name: count, dtype: int64

In [11]:
df.to_csv("../data/cleaned_fatality_data.csv", index=False)
